### ISSUE TITLE
Auto detect site-lisp installation folder

### ISSUE THREAD

**Issue by vspinu** [emacs-ess/ESS#754]
Instead of hard-coding site-lisp location in Makeconfig we could look at the entries of the load-path in `$(EMACS)` -q. This is how path looks on debian-like `"/etc/emacs25" "/etc/emacs" "/usr/local/share/emacs/25.2/site-lisp" "/usr/local/share/emacs/site-lisp"`. 

Judging from the Makeconfig examples for other systems all those paths have "lisp" in it, so we can just pick first existing "lisp" directory from the emacs -q load  path and use it as our SITELISP. The info dir can be picked from `Info-default-directories` and etc dir is placed in `../etc` which would obliviate the fragile etc lookup which we have right now in `ess-utils.el`. 

I think this would reduce the need for path configuration only to dist managers. Ordinary users wouldn't need to bother at all. This would also avoid the common cross compilation issues caused by installing with one emacs and running on another.  

What do you think?

**Comment by jabranham** [comment]
Makes sense to me but wouldn't we want the non-versioned site-lisp folder? Otherwise ESS will disappear if users update their Emacs.

**Comment by vspinu** [comment]
Indeed.

**Comment by vspinu** [comment]
On my machine the unversioned site-lisps don't contain `subdirs.el` which are used to add subdirectories to the load path. As a result the users see only ess-site.el which we simlink explicitely, but not other files, including ess-autoloads.el. So our ess-autoload instructions currently fail with the `make install` case.

**Comment by jabranham** [comment]
I've wondered about that weird debian-only symlink line in lisp/Makefile a few times. I guess this explains it. FWIW my system has subdirs.el in both the versioned and non-versioned site-lisp folders.

**Comment by dickmao** [comment]
I see three ways to install ESS.  Melpa, Debian-RPM, and download tarball from `http://ess.r-project.org/downloads/ess`.  It seems like you are trying to facilitate the third, ornery option, which I'm not sure is worth it.

**Comment by vspinu** [comment]
Not the third, the forth. I think git clone + make install is way more common than tarball download. This proposal is an alternative to your home install idea. Either one or another. The home install would also require auto-detection. I personally don't keep .emacs.d in the default location. 

We also haven't heard comments on this particular feature from @mmaechler, without whose approval we cannot merge such a major change. The home install part of the PR (the elephant) seemed to have been overlooked. Yet again, more incremental changes would have been easier to digest.

**Comment by dickmao** [comment]
> The home install would also require auto-detection. I personally don't keep .emacs.d in the default location.

Okay, but hand-editing 1-2 lines in `Makeconf.local` is preferable to performing a autodetect heuristic.  The esteemed `magit` package also contains a notion of manually edited `Makeconf.local` for local installs.  The auto-detection described herein is traditionally implemented via gnu autotools (Makefile.am, configure, autogen) which I'm not sure is worth it (but is the superior solution).

**Comment by vspinu** [comment]
First of all, it's a bit more than a heuristic, it reliably detects site-lisp directory. Second, many ESS users have little clue about where things should go. Finally, no-one plans to obsolete the Makeconf, we just want to provide sane defaults on all systems.

**Comment by mmaechler** [comment]
@dickmao:  I still don't get your basic problem:  We do have a working setup without *any* hand editing, and we definitely don't want you changing that.

Also it seems to me, you (and others in this thread) are working under completely misleading assumptions. Contrary to your assumption, by far most ESS users do get a version of ESS that comes from our tarball distribution (to Windows via Vincent Goulet's,  to Debian and other Linux distributions who will not propate a non-released version either),... and people in the corporate world where some use S-plus (as you now finally have to admit, .. you did not when I first mentioned it) and other also will not use development versions of software but rather official releases.

Please don't fix something that is not broken -- just because it does not look clean or pure or ... to you.  There many more interesting things to be done with emacs lisp.  Keep off the Makefile tweaking, please!

**Comment by lionel-** [comment]
@vspinu If we have package.el tarballs, installation just becomes a matter of calling `package-install-file` on the tarball, and this will work cross-platform on well-configured Emacsen right? In that case we don't have to auto-detect anything?

**Comment by vspinu** [comment]
For local home directory installs indeed. For site-wise we need something more robust than the current setup. Manual hacking of paths is error prone and doesn't even work as expected on debian derivatives for which our Makeconf provides the default paths (due to the missing subdirs.el in the debian site-lisp folders).
